std/random.d

```diff
@@ -1904,17 +1904,35 @@ unittest
     import std.algorithm;
     foreach(RandomGen; PseudoRngTypes)
     {
-        auto a = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+        auto a = [0, 1, 1, 2, 3];
         auto b = a.dup;
         auto gen = RandomGen(unpredictableSeed);
-        partialShuffle(a, 5, gen);
-        assert(a[5 .. $] == b[5 .. $]);
-        sort(a[0 .. 5]);
-        assert(a[0 .. 5] == b[0 .. 5]);
-        partialShuffle(a, 6);
-        assert(a[6 .. $] == b[6 .. $]);
-        sort(a[0 .. 6]);
-        assert(a[0 .. 6] == b[0 .. 6]);
+
+        // NUM times, pick LEN elements from the array at random.
+        immutable int LEN = 2;
+        immutable int NUM = 750;
+        int[][] chk;
+        foreach(step; 0..NUM)
+        {
+            partialShuffle(a, LEN, gen);
+            chk ~= a[0..LEN].dup;
+        }
+
+        // Check that each possible a[0..LEN] was produced at least once.
+        // For a perfectly random RandomGen, the probability that each
+        // particular combination failed to appear would be at most
+        // 0.95 ^^ NUM which is approximately 1,962e-17.
+        // As long as hardware failure (e.g. bit flip) probability
+        // is higher, we are fine with this unittest.
+        sort(chk);
+        assert(equal(uniq(chk), [       [0,1], [0,2], [0,3],
+                                 [1,0], [1,1], [1,2], [1,3],
+                                 [2,0], [2,1],        [2,3],
+                                 [3,0], [3,1], [3,2],      ]));
+
+        // Check that all the elements are still there.
+        sort(a);
+        assert(equal(a, b));
```

Comment:
Hmm.

I can try to convince you the 10^-17 probability of an error is fine, but that looks like a discussion topic for a newsgroup or other media.

I agree that a fixed seed would be a good compromise.  The test is only a bit less convincing then.

Note that the test can still fail with said probability whenever the list of generators is updated.  For particularly bad PRNGs, the probability can be even higher, but it can be argued that these shouldn't be included in the list, or in Phobos at all.